### PR TITLE
Use already-installed tools in `uv tool run`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This doesn't cache the tool environment; rather, it just uses the `tool install` environment if it satisfies the request.

Closes https://github.com/astral-sh/uv/issues/4742.

### CODE REVIEW

File: crates/uv/tests/tool_run.rs [hunk 1]
```diff
@@ -151,3 +151,62 @@ fn tool_run_from_version() {
...
+    warning: `uv tool run` is experimental and may change without warning.
+    "###);
+
+    // Verify that `tool run black` at a different version installs the new version.
+    uv_snapshot!(context.filters(), context.tool_run().arg("black@23.12.1").arg("--version"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    black, 24.1.0 (compiled: yes)
+    Python (CPython) 3.8.19
```
Comment (zanieb) [inline crates/uv/tests/tool_run.rs:197]: Whoopsie?

File: crates/uv/tests/tool_run.rs [hunk 2]
```diff
@@ -151,3 +151,62 @@ fn tool_run_from_version() {
...
+
+    ----- stderr -----
+    warning: `uv tool run` is experimental and may change without warning.
+    Resolved 6 packages in [TIME]
+    Prepared 1 package in [TIME]
+    Installed 6 packages in [TIME]
+     + black==23.12.[X]
```
Comment (zanieb) [inline crates/uv/tests/tool_run.rs:204]: Funny that it gets installed but not used.

**Comment by zanieb** [comment]
Dangerous to review your pull requests so soon.

We should include tests with:

- compatible and incompatible constraints e.g. `uv tool run --from 'black>23' black`
- `--isolated` to opt-out and use the latest version
- an extra requirement via `--with`

**Comment by charliermarsh** [comment]
Good call, added!

**Review by zanieb — APPROVED** [review]

File: crates/uv/tests/tool_run.rs [hunk 3]
```diff
@@ -151,3 +197,142 @@ fn tool_run_from_version() {
...
+     + click==8.1.7
+     + mypy-extensions==1.0.0
+     + packaging==24.0
+     + pathspec==0.12.1
+     + platformdirs==4.2.0
+    "###);
+
+    // Verify that `--with` installs a new version.
+    // TODO(charlie): This could (in theory) layer the `--with` requirements on top of the existing
+    // environment.
```
Comment (zanieb) [inline crates/uv/tests/tool_run.rs:311]: This would be nice and sensible

**Pull request merged**